Sources/ApolloCodegenLib/Frontend/JavaScriptBridge.swift

```diff
@@ -0,0 +1,345 @@
+  public var description: String {
+    return jsValue.toString()
+  }
+}
+
+/// A type that references an underlying JavaScript object.
+public class JavaScriptObject: JavaScriptValueDecodable {
+  let jsValue: JSValue
+  unowned let bridge: JavaScriptBridge
+  
+  static func fromJSValue(_ jsValue: JSValue, bridge: JavaScriptBridge) -> Self {
+    return bridge.wrap(jsValue)
+  }
+    
+  required init(_ jsValue: JSValue, bridge: JavaScriptBridge) {
+    precondition(jsValue.isObject)
+        
+    self.jsValue = jsValue
+    self.bridge = bridge
+  }
+  
+  subscript(property: Any) -> JSValue {
+    return jsValue[property]
+  }
+  
+  subscript<Decodable: JavaScriptValueDecodable>(property: Any) -> Decodable {
+    return bridge.fromJSValue(jsValue[property])
+  }
+  
+  private func call(_ functionName: String, with arguments: [Any]) throws -> JSValue {
+    return try bridge.throwingJavaScriptErrorIfNeeded {
+      let function = jsValue[functionName]
+      
+      precondition(!function.isUndefined, "Function \(functionName) is undefined")
+      
+      return function.call(withArguments: bridge.unwrap(arguments))!
+    }
+  }
+  
+  func call(_ functionName: String, with arguments: Any...) throws -> JSValue {
+    try call(functionName, with: arguments)
+  }
+  
+  func call<Decodable: JavaScriptValueDecodable>(_ functionName: String, with arguments: Any...) throws -> Decodable {
+    return bridge.fromJSValue(try call(functionName, with: arguments))
+  }
+  
+  private func invokeMethod(_ methodName: String, with arguments: [Any]) throws -> JSValue {
+    return try bridge.throwingJavaScriptErrorIfNeeded {
+      jsValue.invokeMethod(methodName, withArguments: bridge.unwrap(arguments))
+    }
+  }
+  
+  func invokeMethod(_ methodName: String, with arguments: Any...) throws -> JSValue {
+    try invokeMethod(methodName, with: arguments)
+  }
+  
+  func invokeMethod<Decodable: JavaScriptValueDecodable>(_ methodName: String, with arguments: Any...) throws -> Decodable {
+    return bridge.fromJSValue(try invokeMethod(methodName, with: arguments))
+  }
+  
+  func construct<Wrapper: JavaScriptObject>(with arguments: Any...) throws -> Wrapper {
+    return bridge.fromJSValue(try bridge.throwingJavaScriptErrorIfNeeded {
+      jsValue.construct(withArguments: bridge.unwrap(arguments))
+    })
+  }
+}
+
+extension JavaScriptObject: Equatable {
+  public static func ==(lhs: JavaScriptObject, rhs: JavaScriptObject) -> Bool {
+    return lhs.jsValue == rhs.jsValue
+  }
+}
+
+
+/// The JavaScript bridge is responsible for converting values to and from type-safe wrapper objects. It also ensures exceptions thrown from JavaScript wrapped and rethrown.
+class JavaScriptBridge {
+  private var context: JSContext
+  
+  // In JavaScript, classes are represented by constructor functions.
+  // We keep a bidirectional mapping between constructors and wrapper types so we can create wrappers
+  // for JavaScript objects and
```

Comment:
looks like you got distracted in the middle of a thought here - I'm not quite clear on the comments below either. Can you go into a little more detail here?

Reply:
Yep, will do :)